### PR TITLE
Add OpenBMB MiniCPM-V 4.6 support

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

- add native `minicpmv4_6` support for `openbmb/MiniCPM-V-4.6` / Transformers 5.7+
- implement packed variable-resolution SigLIP2 vision, ragged overview/slice batching, inserted window-attention merger, final MLP merger, and 16x/4x modes
- reuse the Qwen3.5 hybrid decoder and standard three-model VLM helpers; support image and video placeholder routing
- add config/registry/CLI/runtime integration, weight mapping, synthetic and real-weight parity, and genuine image-conditioned L4/L5 goldens

## Updated from main

Rebased the linear six-commit PR series onto current `origin/main` at `c3b7cf1`, incorporating Mage-VL image/video support and Parakeet CTC in addition to Moonshine, Nemotron Parse 2.0, LFM2.5, fused Swish, Gemma 3n, and Muse Glimmer. Conflict resolution preserves all upstream registrations, configs, tasks, docs, and tests alongside MiniCPM.

## Validation after latest rebase

- `lintrunner -a`: clean
- config extractor: `1 passed, 24 deselected`
- MiniCPM unit/L1: `6 passed, 1392 deselected`
- CUDA real-weight nonzero-image parity: `1 passed, 126 deselected`
- prior CPU real parity and CUDA L4/L5 image generation remain unchanged by the registration-only main conflicts
- FP16 CUDA stages: vision cosine `0.999961913`, embedding cosine `0.999961913`, last-token logits cosine `0.999991179`, identical argmax `760`, no NaNs

## Fused and optimized graphs

Weighted FP16 CUDA-targeted graph audit after the latest rebase:

| Graph | Nodes | Identity | Unused inputs | Fused/layout evidence |
|---|---:|---:|---:|---|
| decoder | 1,107 | 0 | 0 | 6 GQA, 18 LinearAttention, 18 CausalConvWithState, 48 SkipSimplifiedLayerNormalization, 42 Swish, 36 transpose |
| vision_encoder | 619 | 0 | 0 | 28 Attention, 54 SkipLayerNormalization, 5 transpose, 2 cast |
| embedding | 15 | 0 | 0 | 0 transpose, 1 cast, 2 reshape |

Initializer folding reduces unweighted vision transposes from 175 to five required raster/window permutations. FP16 avoids BF16-only compatibility casts.

Actual CUDA profiling confirms the fused kernels execute on CUDA: vision 542 CUDA nodes / 0 CPU compute, embedding 15 CUDA / 0 CPU, decoder 991 CUDA nodes with all GQA/LinearAttention/CausalConv/skip-RMSNorm kernels on CUDA. Outputs are finite and image-conditioned argmax remains `760`.

## DRYness

Reuses `Qwen35TextModel`, shared hybrid-cache task machinery, `VisionEncoderLayer`, `VisionAttention`, `LayerNorm`, `Linear`, activation helpers, `VisionLanguageTask`, `build_decoder_from_embeds`, and `build_embedding_from_features`. Only MiniCPM-specific NaViT/ragged-grid/merger logic remains local. The specialized dual image/video embedding graph is 15 nodes.

## Runtime limitations

- ort-extensions has no native MiniCPM adaptive slicing + horizontal NaViT packing transform; callers supply Hugging Face processor tensors, including native INT32 `target_sizes`.
- `onnxruntime-genai-cuda 0.15.2` crashes loading this hybrid package on Windows, while raw ORT CUDA executes all three graphs and exact L5 generation successfully.
- FP16 is the validated end-to-end reduced-precision CUDA path; BF16 decoder awaits an ORT `CausalConvWithState` kernel.
- Foundry Local and Olive are not installed in this environment.

### CODE REVIEW

File: src/mobius/tasks/_vision_language_3model.py [hunk 1]
```diff
@@ -250,6 +250,68 @@ def build(
...
+            shape=[
+                packed_batch,
+                vision_config.in_channels,
+                patch_size,
+                packed_width,
+            ],
+        )
+        target_sizes = builder.input(
+            "target_sizes",
+            dtype=ir.DataType.INT32,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/tasks/_vision_language_3model.py:364]: Keeping this INT32 is intentional and matches the native Transformers MiniCPMV4_6ImageProcessor contract: it constructs target_sizes with torch.tensor(..., dtype=torch.int32). The checked-in real processor tests and CUDA parity feeds use that output directly, so changing the graph input to INT64 would introduce the boundary cast rather than remove one. image_grid_thw is a different processor contract.

File: src/mobius/_configs/per_model/_minicpmv4_6_vision.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        image_size=image_size,
+        patch_size=patch_size,
+        norm_eps=vision.layer_norm_eps,
+        hidden_act=getattr(vision, "hidden_act", "gelu_pytorch_tanh"),
+        in_channels=getattr(vision, "num_channels", 3),
+        num_position_embeddings=(image_size // patch_size) ** 2,
+        image_token_id=parent_config.image_token_id,
+        insert_layer_id=getattr(parent_config, "insert_layer_id", 6),
+        window_kernel_size=tuple(getattr(vision, "window_kernel_size", (2, 2))),
+        merge_kernel_size=tuple(getattr(parent_config, "merge_kernel_size", (2, 2))),
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_configs/per_model/_minicpmv4_6_vision.py]: Fixed. Both kernel fields now use getattr(..., None) or (2, 2) before tuple conversion, and a regression test covers explicit None values for both window_kernel_size and merge_kernel_size.

**Comment by justinchuby** [comment]
Looks like we need compress and argmax in ORT cuda

File: src/mobius/models/minicpmv4_6.py [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+        image_features: ir.Value,
+    ):
+        text_embeds = self.embed_tokens(op, input_ids)
+        media_mask = op.Or(
+            op.Equal(input_ids, self.image_token_id),
+            op.Equal(input_ids, self.video_token_id),
+        )
+        media_indices = op.Clip(
+            op.Sub(
+                op.CumSum(op.Cast(media_mask, to=7), 1),
```
Comment (justinchuby) [inline src/mobius/models/minicpmv4_6.py]: `CumSum(..., axis=1)` resets feature indices for every batch row, but `image_features` is one globally packed stream. With two prompts containing distinct media, row 2 gathers again from feature 0 and receives row 1’s visual features. Please compute indices over a batch-major flattened media mask (then reshape), and add a two-row test with distinguishable features.

**Review by justinchuby — COMMENTED** [review]
Correctness review: one batched multimodal finding.

File: src/mobius/models/minicpmv4_6.py (hunk 3 above)
Comment (justinchuby) [reply src/mobius/models/minicpmv4_6.py]: Fixed in f1467bc. The embedding mixer now flattens the batch-major media mask before CumSum and reshapes gathered features back to the token embedding shape. Added a two-row masked_scatter parity test with distinct image/video features plus batched empty decode coverage.

File: src/mobius/models/minicpmv4_6_test.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/minicpmv4_6_test.py:1]: Fixed in 1400b7b by applying the repository formatter to the media embedding test.
